csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -50,1508 +50,1501 @@ private import SuccessorTypes
-    any(AssignOperation ao |
-      result = ao.getExpandedAssignment() and
-      c = TRec(TLastRecAnyCompletion())
-    )
-  or
-  cfe =
-    any(ConditionallyQualifiedExpr cqe |
-      // Post-order: element itself
-      result = cqe and
-      c = getValidSelfCompletion(result)
+    final override predicate propagatesAbnormal(ControlFlowElement child) {
+      child = getExprChild(this, _)
       or
-      // Qualifier exits with a `null` completion
-      result = getExprChildElement(cqe, 0) and
-      c = TRec(TLastRecSpecificCompletion(any(NullnessCompletion nc | nc.isNull())))
-    )
-  or
-  // Expression being thrown exits abnormally
-  result = cfe.(ThrowExpr).getExpr() and
-  c = TRec(TLastRecAbnormalCompletion())
-  or
-  cfe =
-    any(ObjectCreation oc |
-      // Post-order: element itself (when no initializer)
-      result = oc and
-      not oc.hasInitializer() and
-      c = getValidSelfCompletion(result)
+      child = this.getCall(_)
+    }
+
+    final override predicate first(ControlFlowElement first) { first(getExprChild(this, 0), first) }
+
+    final override predicate succ(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
+      // Standard left-to-right evaluation
+      exists(int i |
+        last(getExprChild(this, i), pred, c) and
+        c instanceof NormalCompletion and
+        first(getExprChild(this, i + 1), succ)
+      )
       or
-      // Last element of initializer
-      result = oc.getInitializer() and
-      c = TRec(TLastRecAnyCompletion())
-    )
-  or
-  cfe =
-    any(ArrayCreation ac |
-      // Post-order: element itself (when no initializer)
-      result = ac and
-      not ac.hasInitializer() and
-      c = getValidSelfCompletion(result)
+      // Flow from last element of last child to first accessor call
+      last(getLastExprChild(this), pred, c) and
+      succ = this.getCall(0) and
+      c instanceof NormalCompletion
       or
-      // Last element of initializer
-      result = ac.getInitializer() and
-      c = TRec(TLastRecAnyCompletion())
-    )
-  or
-  cfe =
-    any(IfStmt is |
-      // Condition exits with a false completion and there is no `else` branch
-      result = is.getCondition() and
-      c = specificBoolean(false) and
-      not exists(is.getElse())
+      // Flow from one call to the next
+      exists(int i | pred = this.getCall(i) |
+        succ = this.getCall(i + 1) and
+        c.isValidFor(pred) and
+        c instanceof NormalCompletion
+      )
       or
-      // Condition exits abnormally
-      result = is.getCondition() and
-      c = TRec(TLastRecAbnormalCompletion())
+      // Post-order: flow from last call to element itself
+      exists(int last | last = max(int i | exists(this.getCall(i))) |
+        pred = this.getCall(last) and
+        succ = this and
+        c.isValidFor(pred) and
+        c instanceof NormalCompletion
+      )
+    }
+  }
+
+  private class LogicalNotExprTree extends PostOrderTree, LogicalNotExpr {
```

Comment:
I would have thought before that we don't need any special handling of expressions with a single child. I understand that this is needed because of having `BoolCompletion` that needs to be negated from the child.

But this also means to me that when we have a pattern match, such as `x is object`, this has also has a bool completion, which somehow need to end up as a `NullnessCompletion(false)`. (This is true for type patterns, but not for `VarPatternSyntax`). Yet, there's no special handling of pattern expressions here, why? (There's some for `CaseTree `, but not for `IsExpr`.)

To be fair, I'm not even sure if we end up with a `NullnessCompletion` in this case.

Reply:
This is actually some of what I am working on in my PR for implementing the CFG for `not` patterns. Right now, the CFG for `if (x is object) ...` looks like
```
    x
    |
  object
    |
x is object
   /  \
true  false
  |     |
 ...   ...
```

After the PR it will look like
```
          x
          |
        object
         / \
        /   \
       /     \
      /       \
     /         \
  match      no-match
    |           |
x is object   x is object
    |           |
   true       false
    |           |
   ...         ...
```

Reply:
(For completeness): The CFG for `x is not object` will be
```
              x
              |
            object
             / \
            /   \
           /     \
          /       \
         /         \
      match      no-match
        |           |
    not object   not object
        |           |
     no-match     match
        |           |
 x is not object   x is not object
        |           |
       false       true
        |           |
       ...         ...
```

Reply:
Sounds good. 
If I understand it correctly then the non-nullness information in `x is object` would be basically a `TMatchingCompletion(true)` on a `TypePatternExpr`.

Reply:
Indeed (once the CFG / guards library are updated).